Tracks variable names from in_fmt

sjavis opened the pull request:
TETracker.tracks will now use `in_fmt` (which in turn uses detect nodes' `output_commands` if left blank) for the variable names when reading in gfdl / csvnohead output files. Closes #40.
I have also made `in_fmt` a list rather than requiring a comma-separated-value string.

This doesn't currently use any other parameters because that would be done during the CF file writing.

jatkinson1000 reviewed (commented):
Hi @sjavis This generally looks good, but you'll want to rebase to resolve a couple of merge conflicts.

Other question is if you checked the documentation to see if there is anything that needs updating - it may well be from a quick look that the answer is no given that this is somewhat 'internal'.

I also wonder if we need to look through and be slightly clearer in the docs which options a user should _not_ set if running as part of chained call. Perhaps this is something for #27?

sjavis commented:
Rebased to the latest cf-outputs.

I don't think there are any other places in the docs that need to be updated. The example usage gets the format from the detect nodes parameters. I agree about needing something clearer about what to avoid for a full run (relatedly I have been wondering about temporary output files for DetectNodes / StitchNodes), but I will leave this for #27.

I will leave this for you to merge (if you're happy) since it is into the branch you're working on.

sjavis commented:
I've added new tests to check the assignment of StitchNodesParameters' in_fmt/in_file and to explicitly check the variable names are being set correctly when reading gfdl and csvnohead files. I also took the implicit check for this out of `test_tracks`.

sjavis commented:
Rebased to pick up latest changes in cf-output

jatkinson1000 commented on src/tctrack/tempest_extremes/tempest_extremes.py:
```diff
@@ -968,6 +968,9 @@ def _parse_tracks_gfdl(self, file_path):
         current_track_id = 0  # Initialize track ID
 
+        # Get variable names from in_fmt
+        var_names = self.stitch_nodes_parameters.in_fmt or []
```
I think the below may be 'safer' as `_parse_gfdl_line_to_point` expects a list of variable names or `None` as an input. Whilst `if([])` returns `False` in Python there is perhaps potential for confusion later, whilst doing this would propogate the list of names or `None` directly as set in `self.stitch_nodes_parameters.in_fmt`

```suggestion
        var_names = self.stitch_nodes_parameters.in_fmt
```

sjavis replied:
Good point, I just got confused myself when I looked at what it was currently doing.

jatkinson1000 commented on tests/unit/test_tempest_extremes.py:
```diff
@@ -577,6 +577,33 @@ def test_stitch_nodes_threshold_filters(
         assert result["stderr"] == "Mocked stderr output"
         assert result["returncode"] == 0
 
+    def test_stitch_nodes_values_from_dn(self) -> None:
```
Consistent naming in function name, and explicit.
```suggestion
    def test_stitch_nodes_values_from_detect_nodes(self) -> None:
```

jatkinson1000 commented on tests/unit/test_tempest_extremes.py:
```diff
@@ -577,6 +577,33 @@ def test_stitch_nodes_threshold_filters(
+            output_file="file.txt", output_commands=output_commands
+        )
+
+        # Check defaults
+        tracker = TETracker()
+        assert tracker.stitch_nodes_parameters.in_file is None
+        assert tracker.stitch_nodes_parameters.in_fmt is None
+
+        # Check the auto-assignment works
+        tracker = TETracker(dn_params)
+        assert tracker.stitch_nodes_parameters.in_file == "file.txt"
+        assert tracker.stitch_nodes_parameters.in_fmt == ["lon", "lat", "v1", "v2"]
+
+        # Ensure defined values don't get overriden
+        sn_params = StitchNodesParameters(in_file="file2.txt", in_fmt=["a", "b"])
+        tracker = TETracker(dn_params, sn_params)
+        assert tracker.stitch_nodes_parameters.in_file == "file2.txt"
+        assert tracker.stitch_nodes_parameters.in_fmt == ["a", "b"]
+
```
I think this is OK, but really we are testing 3 edge cases here so would prefer to see this as a test parameterised with `dn_params`, `sn_params`, and expected outputs.
However, given time constraints on the project I don't think this is a blocker for this PR.

jatkinson1000 replied:
Thanks, would have been OK without, but definitely an improvement longer term, so thanks for adding this as well!

jatkinson1000 commented on tests/unit/test_tempest_extremes.py:
```diff
@@ -761,3 +788,31 @@ def test_tracks(self, file_format, mock_file_fixture, expected_tracks, request):
+            ("csvnohead", "mock_csvnohead_file"),
+        ],
+    )
+    def test_tracks_header_names(self, file_format, mock_file_fixture, request) -> None:
+        """Test the track header names are assigned correctly by tracks()."""
+        # Use DetectNodesParameters to set the variable names
+        output_commands = [
+            TEOutputCommand(var="v1", operator="min", dist=0.0),
+            TEOutputCommand(var="v2", operator="min", dist=0.0),
+        ]
```
Yes, this is a good way of checking this I think.

jatkinson1000 reviewed (commented):
@sjavis I think this generally looks OK, thanks for revisiting the tests after my review.

If you can read the comments and accept the suggestions if you are happy with them then I'll look at approving and merging this.

sjavis commented:
I have addressed the comments, mainly modifying the test for the assignment of the stitch nodes parameters. If you are happy you can go ahead with the merge.
